lib/haml/parser.rb

```diff
@@ -698,7 +698,7 @@ def parse_new_attributes(text)
       end
 
       static_attributes = {}
-      dynamic_attributes = "{"
+      dynamic_attributes = ["{"]
```

Comment:
Doesn't impact performance or memory either way but is necessary for future `future_string_literal: true` changes

```ruby
# frozen_string_literal: false
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update
                your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  gem "benchmark-ips"
  gem "rails"
end

def allocate_count
  GC.disable
  before = ObjectSpace.count_objects
  yield
  after = ObjectSpace.count_objects
  after.each { |k,v| after[k] = v - before[k] }
  after[:T_HASH] -= 1 # probe effect - we created the before hash.
  GC.enable
  result = after.reject { |k,v| v == 0 }
  GC.start
  result
end

def master_version
  dynamic_attributes = '{'
  if rand(2) == 1
    dynamic_attributes << 'hi there'.freeze
  end
  dynamic_attributes << "}".freeze
  dynamic_attributes = nil if dynamic_attributes == "{}".freeze
end

def fast_version
  dynamic_attributes = ['{'.freeze]
  if rand(2) == 1
    dynamic_attributes << 'hi there'.freeze
  end
  dynamic_attributes << "}".freeze
  dynamic_attributes = nil if dynamic_attributes.first == "{}".freeze && dynamic_attributes.length == 1
end

puts "master_version"
puts allocate_count { 1000.times { master_version } }
puts "fast_version"
puts allocate_count { 1000.times { fast_version } }

Benchmark.ips do |x|
  x.report("master_version") { master_version }
  x.report("fast_version")     { fast_version }
  x.compare!
end
```
```
master_version
{:FREE=>-693, :T_STRING=>1052, :T_IMEMO=>1}
fast_version
{:FREE=>-1001, :T_ARRAY=>1000}
Warming up --------------------------------------
      master_version   111.400k i/100ms
        fast_version   122.259k i/100ms
Calculating -------------------------------------
      master_version      1.976M (±17.1%) i/s -      9.580M in   5.056250s
        fast_version      1.966M (±25.6%) i/s -      8.925M in   5.017136s

Comparison:
      master_version:  1975829.5 i/s
        fast_version:  1965749.8 i/s - same-ish: difference falls within error
```

Reply:
Thanks for your effort of gradual changes and to put benchmark, but please exclude intermediate changes that have no benefit. It makes hard to read commit log to know why this change was necessary, and this change should be committed with `# frozen_string_literal: true`.

Reply:
And there's no need to change this to Array buffer for introducing `# frozen_string_literal: true`.

It uglifies code after this, so please use `"{".dup` instead.